### PR TITLE
roachprod: enable roachprod to be configured for external use

### PR DESCRIPTION
**Pull request by ajwerner**
After this change, roachprod fully works on GCP at my company with the following configuration:

```
ROACHPROD_DNS_REQUIRED_PROVIDERS=gce
ROACHPROD_EMAIL_DOMAIN=@dataexmachina.dev
ROACHPROD_GCE_DEFAULT_PROJECT=dem-ephemeral
ROACHPROD_GCE_DEFAULT_SERVICE_ACCOUNT=roachprod@dem-ephemeral.iam.gserviceaccount.com
ROACHPROD_GCE_DNS_DOMAIN=roachprod.exray.cloud
ROACHPROD_GCE_DNS_MANAGED_DOMAIN=roachprod-managed.dataexmachina.dev
ROACHPROD_GCE_DNS_MANAGED_ZONE=roachprod-managed
ROACHPROD_GCE_DNS_PROJECT=dem-ephemeral
ROACHPROD_GCE_DNS_ZONE=roachprod
```

All of these configuration options can also be set via flags.

See individual commits.

Epic: none
Release note: none

### CODE REVIEW

File: pkg/cmd/roachprod/flags.go
Comment (srosenberg) [inline pkg/cmd/roachprod/flags.go:136]: Methinks only `setupSSHCmd` calls `Sync`, which in turn may invoke `SyncDNS`. The others don't need parameterized DNS flags.

**Review by srosenberg — APPROVED** [review]
Thanks for refactoring hardcoded configs and making `roachprod` useful externally! I played around with the PR locally; things appear to work as they should. Still, it wouldn't hurt to do a full run of the nightlies as a sanity check.

@renatolabs @herkolategan I'd appreciate another pair of eyes.

**Comment by andreimatei** [comment]
I rebased the patches. A review would be greatly appreciated; this PR will help us do other good things for cockroach :)
In particular, I want to propose some `roachtest` patches and getting `roachprod` working locally in an easier way would be great.

**Comment by srosenberg** [comment]
> I rebased the patches. A review would be greatly appreciated; this PR will help us do other good things for cockroach :) In particular, I want to propose some `roachtest` patches and getting `roachprod` working locally in an easier way would be great.

Thanks and sorry for the delay! I am happy with the changes. The reason it probably stalled is we didn't do any smoke testing. Lemme do some local testing as well as kick off a nightly run with `SELECT_PROBABILITY`.

**Comment by srosenberg** [comment]
Will check the smoke test, [SELECT_PROBABILITY=0.6](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/15369312?buildTab=log&logFilter=debug&focusLine=5&logView=linear), when it completes.

**Comment by andreimatei** [comment]
Ugh, I think there might be something broken, so please hold on on the testing. I think things are good when `roachprod` is used as a binary, but not when `roachtest` uses it as a library. I'm working on `roachtest` myself, so I noticed.
I'll ping back when I've figured it out. Sorry for the noise.

**Comment by srosenberg** [comment]
> Ugh, I think there might be something broken, so please hold on on the testing. I think things are good when `roachprod` is used as a binary, but not when `roachtest` uses it as a library. I'm working on `roachtest` myself, so I noticed. I'll ping back when I've figured it out. Sorry for the noise.

No worries! Yep, seeing a lot of these in CI, 

```
ERROR: (gcloud.compute.project-info.describe) The project property is set to the empty string, which is invalid.
```

**Comment by andreimatei** [comment]
OK, I believe I've fixed the issue. roachtests run fine for me now.
I've also had to pile on one more commit to optionally disable the integration with the Prometheus registration service. PTAL
Thanks!

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> Reviewed 5 of 6 files at r6, 2 of 3 files at r8, 5 of 5 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120340#-:-Nyb7l7M2hg1QA7IgxgS:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @DarrylWong, @herkolategan, and @renatolabs)

**Comment by srosenberg** [comment]
Smoke test: [SELECT_PROBABILITY=0.6](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/15383971?buildTab=log&logFilter=debug&focusLine=5&logView=linear).

**Comment by andreimatei** [comment]
I've pushed the fix to the problem I was describing in https://github.com/cockroachdb/cockroach/pull/124678#issuecomment-2136040318.
If you kick off another smoke test, I'll verify in the runner logs that the cluster reuse is working. (a more limited run of only a few tests, for example `roachtest run acceptance`, should do if that's easy).
Otherwise, the previous run looked fine, right?
Thanks!

**Comment by srosenberg** [comment]
Kicked off another run of [SELECT_PROBABILITY=0.6](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/15443185?buildTab=log&logFilter=debug&focusLine=6&logView=linear).

**Comment by andreimatei** [comment]
Cluster reuse looks good now -- you see `Using existing cluster` in the runner's logs.

**Review by herkolategan — APPROVED** [review]

**Comment by srosenberg** [comment]
Smoke test results look good! Resolving merge conflict and merging. Thanks for making roachprod ready for external use!

**Comment by srosenberg** [comment]
Final smoke test ([SELECT_PROBABILITY=0.25](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/15462087?buildTab=log&logFilter=debug&focusLine=6&logView=linear&linesState=3) since there were non-trivial merge conflicts after   [1] got merged.

[1] https://github.com/cockroachdb/cockroach/pull/124604

**Comment by srosenberg** [comment]
Final smoke test looks good, let's 🚢 

bors r+

**Pull request merged**
